### PR TITLE
Feat: Use sandbox id by default for generation commands

### PR DESCRIPTION
**Pull request by sdstolworthy**
*Issue #, if available:*

*Description of changes:*
See https://github.com/aws-amplify/amplify-backend/pull/605 for previous comments.

This commit genericizes the Fallback resolver somewhat, accepting a default resolver and a fallback resolver. 

Additionally, `BackendIdentifierResolver['resolve']` can now return undefined, addressing the comment thread here: https://github.com/aws-amplify/amplify-backend/pull/605#discussion_r1387296389.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by sobolk — DISMISSED** [review]

File: packages/cli/src/backend-identifier/backend_identifier_resolver.ts [hunk 1]
```diff
@@ -6,6 +6,13 @@ type BackendIdentifierParameters = {
   appId?: string;
   branch?: string;
 };
+
+export type BackendIdentityResolver = {
```
Comment (edwardfoyle) [inline packages/cli/src/backend-identifier/backend_identifier_resolver.ts:10]: Having `BackendIdentifierResolver` and `BackendIdentityResolver` is confusing. It's difficult to know what the different between each is. Could we have a single `BackendIdentifierResolver` type with two implementation: one that resolves a single type of BackendId and another that goes through the fallback chain?

If that proves to be too much refactoring, I think I'd rather just add the sandbox fallback directly to the existing `BackendIdentifierResolver`

File: packages/cli/src/backend-identifier/backend_identifier_with_sandbox_fallback.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import { SandboxBackendIdResolver } from '../commands/sandbox/sandbox_id_resolver.js';
+import { BackendIdentifierResolver } from './backend_identifier_resolver.js';
+
+/**
+ * Resolves the backend id when branch or stack is passed as an arg, otherwise returns a sandbox backend identifier
+ */
+export class BackendIdentifierResolverWithFallback {
```
Comment (edwardfoyle) [inline packages/cli/src/backend-identifier/backend_identifier_with_sandbox_fallback.ts:7]: should this implement `BackendIdentityResolver`?

File: packages/cli/src/backend-identifier/backend_identifier_resolver.ts (hunk 1 above)
Comment (sdstolworthy) [reply packages/cli/src/backend-identifier/backend_identifier_resolver.ts]: @edwardfoyle How do you feel about this? https://github.com/aws-amplify/amplify-backend/pull/669/commits/2dc8035161bada79f18f71ae6841e23d08069e3c#diff-9db562c416fe5dbc758bce7b582f0b1c080d36a408d4e2b12c457f9bbc7645b3R20

File: packages/cli/src/backend-identifier/backend_identifier_with_sandbox_fallback.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import { SandboxBackendIdResolver } from '../commands/sandbox/sandbox_id_resolver.js';
+import { BackendIdentifierResolver } from './backend_identifier_resolver.js';
+
+/**
+ * Resolves the backend id when branch or stack is passed as an arg, otherwise returns a sandbox backend identifier
+ */
+export class BackendIdentifierResolverWithFallback
```
Comment (edwardfoyle) [inline packages/cli/src/backend-identifier/backend_identifier_with_sandbox_fallback.ts:7]: nit:
```suggestion
export class BackendIdentifierResolverChain
```
Could even generalize to taking in an array of resolvers and return the first one that doesn't return undefined. But that could also be a load of YAGNI

File: packages/cli/src/backend-identifier/backend_identifier_with_sandbox_fallback.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+   */
+  constructor(
+    private defaultResolver: BackendIdentifierResolver,
+    private fallbackResolver: SandboxBackendIdResolver
+  ) {}
+  /**
+   * resolves the backend id, falling back to the sandbox id if there is an error
+   */
+  public resolve = async (
+    ...args: Parameters<BackendIdentifierResolver['resolve']>
```
Comment (edwardfoyle) [inline packages/cli/src/backend-identifier/backend_identifier_with_sandbox_fallback.ts:21]: pretty sure this type will be inferred from the implementing interface. If not, just type it with `BackendIdentifierParameters`

File: packages/cli/src/backend-identifier/backend_identifier_with_sandbox_fallback.ts (hunk 2 above)
Comment (sdstolworthy) [reply packages/cli/src/backend-identifier/backend_identifier_with_sandbox_fallback.ts]: see: https://github.com/aws-amplify/amplify-backend/pull/669#discussion_r1396015448

File: packages/cli/src/backend-identifier/backend_identifier_with_sandbox_fallback.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+   * Accepts the sandbox id resolver as a parameter
+   */
+  constructor(
+    private defaultResolver: BackendIdentifierResolver,
+    private fallbackResolver: SandboxBackendIdResolver
+  ) {}
+  /**
+   * resolves the backend id, falling back to the sandbox id if there is an error
+   */
+  public resolve = async (
```
Comment (edwardfoyle) [inline packages/cli/src/backend-identifier/backend_identifier_with_sandbox_fallback.ts:20]: nit: public is the default

**Review by edwardfoyle — DISMISSED** [review]
Nice, much clearer! A few nits if you're feeling generous but no blockers

File: packages/cli/src/backend-identifier/backend_identifier_with_sandbox_fallback.ts (hunk 3 above)
Comment (sdstolworthy) [reply packages/cli/src/backend-identifier/backend_identifier_with_sandbox_fallback.ts:7]: I went down that road last night and decided it was YAGNI and walked it back

File: packages/cli/src/backend-identifier/backend_identifier_with_sandbox_fallback.ts (hunk 4 above)
Comment (sdstolworthy) [reply packages/cli/src/backend-identifier/backend_identifier_with_sandbox_fallback.ts:21]: Ah yep now that this is implementing the interface, this should be removed.

**Review by sobolk — APPROVED** [review]

**Review by edwardfoyle — APPROVED** [review]

**Pull request merged**